src/Mvc/Mvc.Razor/src/ApplicationParts/RazorCompiledItemFeatureProvider.cs

```diff
@@ -36,10 +35,53 @@ public void PopulateFeature(IEnumerable<ApplicationPart> parts, ViewsFeature fea
 
                 foreach (var item in provider.CompiledItems)
                 {
-                    var descriptor = new CompiledViewDescriptor(item);
+                    var compiledItem = item;
+                    if (_hotReloadedViews is not null && _hotReloadedViews.TryGetValue(item.Identifier, out var hotReloadedType))
+                    {
+                        // Determine if a hot reload update is available for this view.
+                        compiledItem = new HotReloadRazorCompiledItem(item, hotReloadedType);
+                    }
+
+                    var descriptor = new CompiledViewDescriptor(compiledItem);
                     feature.ViewDescriptors.Add(descriptor);
                 }
             }
         }
+
+        public void UpdateCache(Type[]? types)
+        {
+            if (types is null)
+            {
+                return;
+            }
+
+            foreach (var type in types)
+            {
+                // The Razor file has a [RazorCompiledItemMetadata("Identifier", "/Index.cshtml")]. We'll look it up.
+                var metadataAttribute = type.GetCustomAttributes<RazorCompiledItemMetadataAttribute>()
```

Comment:
Turns out there was already a type-level attribute in Razor.Hosting that we could use. It's slightly stringly typed (you have to look up an attribute by name), but we already do this when looking up routing metadata in Razor Pages.